### ISSUE TITLE
Rethinking decisions about how we organize the apps repository

### ISSUE THREAD

**Issue by larsks** [operate-first/SRE#369]
Let's say I'd like to import an application (like `metallb` or `kfdefs` or `external-secrets` or whatever) into the apps repository. That application has:

- namespaced resources that we place in a top-level application
- cluster-wide resources (rbacs, crds, etc) that get merged into the `cluster-scope` application

By separating out application components in the `cluster-scope` directory, we introduce two related problems:

- At some point in the future, a new version of that application is released. There are new cluster scoped resources, some of the old ones have been removed, and some have simply been renamed. Because we have dumped the manifests alongside all our other cluster scoped resources, we can't simply "update" it. How do we validate that we have removed and/or renamed everything to match the new configuration?

- We later decide we don't to use the application anymore. Disabling the namespaced resources is simple: just remove the application pointed at the top-level directory. Removing the cluster-scoped resources is *hard*: did we get them all? Did we erroneously delete something that belonged to another application because it had a similar name?

While I understand some of the rational that went into the current layout, I worry that we've created an unnecessary administrative burden that is going to lead to a variety of errors -- at best, failing ci because something is missing or misnamed; at worst, a broken application because critical components simply weren't deployed.

This separation would make sense if we were deploying the cluster-scoped resources of an application for a third part, with the namespaced manifests coming from a different repository. That would allow the third part to manage their namespaced resources without having access to make cluster-wide changes. But when everything is here in the same repository, and we are reviewing every commit, I'm not sure the cost is worth it.

**Comment by HumairAK** [comment]
What I like about our current set up is that it's easy to see what cluster/privileged resources we have deployed as it's all in one location. It's also really consistent, in that, all clusterrolebindings, clusterroles, etc. are found in their respective locations.

It does bother me that in certain cases we are breaking up an application/operator/service into various different locations. And after using our current method, I feel like the cons outweigh the good. One big issue I have, is if I want to add the same application to a different cluster, I have to touch multiple different locations, whereas -- if it was all under one folder per app/service/operator, it would be a lot easier. 

Question: Suppose we separate these, pull them out of `cluster-scope` into their own respective folders, where the rest of their namespacescoped resources live (e.g. for `metalb` the cluster resources would go [here](https://github.com/operate-first/apps/tree/master/metallb)), then how do we define the purpose of `cluster-scope`? Right now I would say, `cluster-scope` is where we put all our cluster/privileged resources. But now that seem like it would need further refinedment...like "all standalone cluster/privileged resources". Is this confusing to a newcomer?  ¯\_(ツ)_/¯

**Comment by 4n4nd** [comment]
My suggested solution is that we keep these specific cluster-scoped resources where they are, but we remove them from the cluster-scope kustomizations and add them to the app specific kustomizations. This way we can keep track of all the manifests related to an app in one kustomization. In future if we remove/update an app, we will know which CRD to remove/update since it will be listed in the app's kustomization.

**Comment by HumairAK** [comment]
I worry that might make things even more confusing. We'll then have manifests in cluster-scope that are not part of the `cluster-resources` app, but are referenced to other apps... the back and forth dependency to me seems confusing/non-intuitive.

**Comment by HumairAK** [comment]
@tumido what do you think?

**Comment by larsks** [comment]
What about bundling resources together as "packages"?  So instead of this in the `cluster-scope/overlays/prod/moc/zero/kustomization.yaml` to enable `metallb`:

```
resources:
  - ../../../../base/core/namespaces/metallb-system
  - ../../../../base/rbac.authorization.k8s.io/clusterrolebindings/metallb-allow-privileged
  - ../../../../base/rbac.authorization.k8s.io/clusterrolebindings/metallb-system:controller
  - ../../../../base/rbac.authorization.k8s.io/clusterrolebindings/metallb-system:speaker
  - ../../../../base/rbac.authorization.k8s.io/clusterroles/metallb-system:controller
  - ../../../../base/rbac.authorization.k8s.io/clusterroles/metallb-system:speaker
```

We would have:

```
resources:
  - ../../../../packages/metallb
```

Where `cluster-scope/packages/metallb/kustomization.yaml` would look like:

```
resources:
  - ../../base/core/namespaces/metallb-system
  - ../../base/rbac.authorization.k8s.io/clusterrolebindings/metallb-allow-privileged
  - ../../base/rbac.authorization.k8s.io/clusterrolebindings/metallb-system:controller
  - ../../base/rbac.authorization.k8s.io/clusterrolebindings/metallb-system:speaker
  - ../../base/rbac.authorization.k8s.io/clusterroles/metallb-system:controller
  - ../../base/rbac.authorization.k8s.io/clusterroles/metallb-system:speaker
```

That preserves the existing organization of files, but at the same time makes it easier to enable/disable something in a particular overlay.

**Comment by HumairAK** [comment]
I like this idea, it seems very straightforward and intuitive.

**Comment by tumido** [comment]
I like the proposed idea. The base stays the same, organized and structured, while we allow the overlays to be simplified and easier to read. I like that, lets try it. :slightly_smiling_face:

**Comment by HumairAK** [comment]
I'm assuming we're all still on board with this

**Comment by larsks** [comment]
You can see how something like this look in process in the [moc-apps](https://github.com/CCI-MOC/moc-apps/) repository, specifically in our [bundles](https://github.com/CCI-MOC/moc-apps/tree/master/cluster-scope/bundles) directory and used in [an overlay](https://github.com/CCI-MOC/moc-apps/blob/master/cluster-scope/overlays/ocp-prod/kustomization.yaml).

**Comment by larsks** [comment]
Closed in favor of #442.
